include/jemalloc/internal/jemalloc_internal.h.in

```diff
@@ -817,6 +832,57 @@ sa2u(size_t size, size_t alignment) {
 	return usize;
 }
 
+JEMALLOC_ALWAYS_INLINE malloc_cpuid_t
+malloc_getcpu(void)
+{
+	assert(have_getcpu);
+#if defined(JEMALLOC_HAVE_SCHED_GETCPU)
+	return (malloc_cpuid_t)sched_getcpu();
+#else
+	return -1;
+#endif
+}
+
+/* Return the chosen arena index based on current cpu. */
+JEMALLOC_ALWAYS_INLINE unsigned
+percpu_arena_choose(void)
+{
+	unsigned arena_ind;
+	assert(have_getcpu && (percpu_arena_mode != percpu_arena_disabled));
+
+	malloc_cpuid_t cpuid = malloc_getcpu();
+	assert(cpuid >= 0);
+	if ((percpu_arena_mode == percpu_arena) ||
+	    ((unsigned)cpuid < ncpus / 2)) {
+		arena_ind = cpuid;
+	} else {
+		assert(percpu_arena_mode == per_phycpu_arena);
+		/* Hyper threads on the same physical CPU share arena. */
+		arena_ind = cpuid - ncpus / 2;
+	}
+
+	return arena_ind;
+}
+
+/* arenas[0...max_ind] is the automatic range. */
+JEMALLOC_ALWAYS_INLINE unsigned
+percpu_arena_max_ind(void)
+{
+	assert(have_getcpu && (percpu_arena_mode != percpu_arena_disabled));
+	if (percpu_arena_mode == per_phycpu_arena && ncpus > 1) {
+		if (ncpus % 2) {
+			/* This likely means a misconfig. */
+			return ncpus / 2;
+		}
+		return ncpus / 2 - 1;
+	} else {
+		return ncpus - 1;
+	}
+}
+
+JEMALLOC_INLINE unsigned arena_ind_get(const arena_t *arena);
```

Comment:
If I understand correctly, you're just suggesting reordering some functions in jemalloc_internal.h.in, which sounds good to me.